packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +401,36 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function isVoidLikeType(typeNode?: TSESTree.TSTypeAnnotation): boolean {
+      const services = getParserServices(context);
+      return (
+        !!typeNode &&
+        !tsutils.isTypeFlagSet(
+          getConstrainedTypeAtLocation(services, typeNode),
+          ts.TypeFlags.VoidLike,
+        )
+      );
+    }
+
+    function targetNodeFunctionAncestorNodeHasVoidReturnType(
+      node: TSESTree.Node,
+    ): boolean {
+      if (!node.parent) {
+        return false;
+      }
+
+      if (
+        AST_NODE_TYPES.FunctionDeclaration === node.parent.type ||
+        AST_NODE_TYPES.ArrowFunctionExpression === node.parent.type ||
+        AST_NODE_TYPES.FunctionExpression === node.parent.type
+      ) {
+        return (
+          !!node.parent.returnType && isVoidLikeType(node.parent.returnType)
```

Comment:
Considering the following case, perhaps we should use signature to check the return type?

[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.3.3&showAST=types&fileType=.tsx&code=C4TwDgpgBAYg9nKBeKAKAlMgfFAbnASwBMBuAKAGM4A7AZ2CmAnoC5YFk1Mkcq64ANhAB0AuAHNUAcgBmCKehJQA9MqgQAThrga0AfnSUa9Rs2AAmThjb5i2KH1qCRYybPmKVa6ok3aNqHAA1uhAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANqRE0aB2iQANJnDYclYZ0UA1SVICSTa9NX44iYxAC%2BAXWMeQHoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
```ts
type Foo = () => void;
const test: Foo = () => console.log('foo'); // error (?)
const test2 = (): void => console.log('foo'); // no error(ok)
```

Reply:
If the function type that returns void is specified, it seems good to receive the effect of the new option!

I edit the code to pass the four test cases below, including the code provided as an example.

```javascript
type Foo = () => void;

// case 1 (ArrowFunction + Signature)
const test: Foo = () => console.log('foo'); // error

// case 2 (ArrowFunction + noSignature)
const test2: Foo = function () { return console.log('foo') } // error

// case 3 (functionExpression + Signature)
const test3: () => void = () => console.log('foo'); // error 

// case 4 (functionExpression + noSignature)
const test4: () => void = function () { return console.log('foo') } // error 
```

Reply:
Hi @auvred, :) 
I thought [the review mentioned in the issue](https://github.com/typescript-eslint/typescript-eslint/pull/8375#pullrequestreview-1897027584) meant [these kind of cases](https://github.com/typescript-eslint/typescript-eslint/pull/8375/files#diff-0acfb0bea43769266683f8645d9804098eeb04529992352922eb00ff5b130334R334).
I didn't just mean typed arrow, I thought it was meaningful to check if the function's signature returns void. Thanks 👍 . I agree to wait for a discussion.

```ts
interface Rule {
  SomeNode(): void;
}

declare const context: {
  report(): void
}

const rule: Rule = {
  SomeNode() {
    return context.report();
  }
}
```

Reply:
@auvred As discussed in the issue, it seems like a good idea for me to start working on this. 

I plan to write code that satisfies the test cases written in this comment, but are there any test cases you would like to add?

```javascript
type Foo = () => void;

// case 1 (ArrowFunction + Signature)
const test: Foo = () => console.log('foo'); // error

// case 2 (ArrowFunction + noSignature)
const test2: Foo = function () { return console.log('foo') } // error

// case 3 (functionExpression + Signature)
const test3: () => void = () => console.log('foo'); // error 

// case 4 (functionExpression + noSignature)
const test4: () => void = function () { return console.log('foo') } // error

// case 5 (functionParameter is Function)
declare function foo(arg: () => void): void;
foo(() => console.log());

// case 6 (HigherOrderFunction)
type HigherOrderType = () => () => () => void;
const x: HigherOrderType = () => () => () => console.log();

// case 7 (interface with Function)
interface Foo {
  foo: () => void
}
function bar(): Foo {
  return {
    foo: () => console.log(),
  };
}

// case 8 (objectMethod)
const q = {
  foo: { bar: () => console.log() },
} as {
  foo: { bar: () => void }
}

```

Reply:
I took those examples from `explicit-function-return-type`'s tests: https://github.com/typescript-eslint/typescript-eslint/blob/ecae04f3800fae255db102055e49eef51e31e006/packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts

You can search for `allowTypedFunctionExpressions` here to find some other cases

> but are there any test cases to add?

After a quick look, I think it would be nice to add several tests with JSX (just like in `explicit-function-return-type.test.ts` mentioned above)